File: src/services/transcriptionService.ts
import { Recording, Transcript } from '../types';

export class TranscriptionService {
  private getApiKey(): string {
    return import.meta.env.VITE_GEMINI_API_KEY || '';
  }

  private async blobToBase64(blob: Blob): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onloadend = () => {
        const base64String = (reader.result as string).split(',')[1];
        resolve(base64String);
      };
      reader.onerror = reject;
      reader.readAsDataURL(blob);
    });
  }

  /**
   * Strictly converts verified lecture audio into a faithful text transcript.
   * NEVER generates study notes, summaries, or exam questions.
   */
  async transcribeRecording(
    recording: Recording,
    audioBlob: Blob,
    onProgress?: (msg: string) => void
  ): Promise<Transcript> {
    // 1. Audio Validation
    if (!audioBlob || audioBlob.size === 0) {
      throw new Error('Audio validation failed: Recording contains 0 bytes of audio.');
    }
    if (recording.durationSeconds <= 0) {
      throw new Error('Audio validation failed: Recording duration is invalid (0 seconds).');
    }

    const apiKey = this.getApiKey();
    if (!apiKey) {
      throw new Error('Gemini API key is not configured. Please set VITE_GEMINI_API_KEY in .env.');
    }

    if (onProgress) onProgress('Encoding audio stream for Gemini Speech Transcription...');
    const base64Audio = await this.blobToBase64(audioBlob);
    const mimeType = audioBlob.type || recording.audioMimeType || 'audio/webm';

    if (onProgress) onProgress('Transcribing spoken lecture audio to verbatim text...');

    const systemPrompt = `You are a faithful speech-to-text transcription engine for academic university lectures.

Your task:
- Convert the spoken audio recording into an accurate, complete, verbatim text transcript.
- Preserve all technical terminology, mathematical terms, proper names, numbers, definitions, and spoken examples.
- Use natural sentence punctuation and paragraph breaks.
- If audio is unclear, output "[inaudible]".
- DO NOT generate study notes, summaries, bullet points, study guides, or exam questions.
- Output ONLY the raw spoken text transcript.`;

    const endpoint = `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key=${apiKey}`;

    const payload = {
      systemInstruction: {
        parts: [{ text: systemPrompt }]
      },
      contents: [
        {
          parts: [
            { text: `Please transcribe the audio from this lecture on "${recording.subject} - ${recording.title}". Output only the verbatim spoken transcript.` },
            {
              inlineData: {
                mimeType: mimeType,
                data: base64Audio
              }
            }
          ]
        }
      ],
      generationConfig: {
        temperature: 0.1
      }
    };

    const response = await fetch(endpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });

    if (!response.ok) {
      const errBody = await response.text();
      console.error('Transcription API error:', response.status, errBody);
      throw new Error(`Transcription service failed (HTTP ${response.status}). Please check network connection and API key.`);
    }

    const data = await response.json();
    const transcriptText = data?.candidates?.[0]?.content?.parts?.[0]?.text;

    // 2. Transcript Validation
    if (!transcriptText || transcriptText.trim().length === 0) {
      throw new Error('Transcription completed but produced empty text. The lecture audio may be silent or corrupted.');
    }

    const transcript: Transcript = {
      id: `tr_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`,
      recordingId: recording.id,
      text: transcriptText.trim(),
      language: 'en',
      durationSeconds: recording.durationSeconds,
      createdAt: Date.now(),
      status: 'COMPLETED'
    };

    return transcript;
  }
}

export const transcriptionService = new TranscriptionService();

File: app/src/main/java/com/example/data/service/TranscriptionService.kt
package com.example.data.service

import android.util.Base64
import com.example.api.Content
import com.example.api.GeminiClient
import com.example.api.GenerateContentRequest
import com.example.api.GenerationConfig
import com.example.api.InlineData
import com.example.api.Part
import com.example.data.models.Recording
import com.example.data.models.Transcript
import com.example.data.models.TranscriptStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID

class TranscriptionService {

    /**
     * Stage 1: Converts genuine audio recording into verbatim transcript text.
     * Guaranteed: Fails explicitly if audio is missing, corrupt, or yields no speech.
     */
    suspend fun transcribeRecording(
        recording: Recording,
        audioFile: File?,
        onProgress: (String) -> Unit = {}
    ): Transcript = withContext(Dispatchers.IO) {
        // 1. Audio validation
        onProgress("Validating audio recording file...")
        if (audioFile == null || !audioFile.exists() || audioFile.length() == 0L) {
            throw IllegalStateException("Cannot transcribe: Audio file is missing or empty (0 bytes).")
        }

        if (recording.durationSeconds <= 0L) {
            throw IllegalStateException("Cannot transcribe: Recording duration is 0 seconds.")
        }

        val apiKey = GeminiClient.getApiKey()
        if (apiKey.isBlank()) {
            throw IllegalStateException("Gemini API key is missing. Please set GEMINI_API_KEY in .env.")
        }

        onProgress("Uploading audio bytes to Gemini 2.0 Flash for speech transcription...")
        val audioBytes = audioFile.readBytes()
        val mimeType = if (audioFile.name.endsWith(".mp3", true)) "audio/mp3" else "audio/mp4"
        val base64Audio = Base64.encodeToString(audioBytes, Base64.NO_WRAP)

        val systemPrompt = """
            You are a professional verbatim speech-to-text transcriber for academic university lectures.
            Your task is to transcribe the provided audio recording into accurate, verbatim text.
            
            Strict Transcription Rules:
            1. Transcribe ONLY the words spoken in the audio recording.
            2. Do not invent, hallucinate, assume, or summarize content not present in the audio.
            3. Preserve technical terms, formulas, course names, and exam warnings exactly as spoken.
            4. If speech is unintelligible at moments, transcribe as best as possible without fabricating content.
            5. Return the raw transcript as plain text. Do not wrap in JSON or Markdown blocks.
        """.trimIndent()

        val request = GenerateContentRequest(
            systemInstruction = Content(parts = listOf(Part(text = systemPrompt))),
            contents = listOf(
                Content(
                    parts = listOf(
                        Part(text = "Transcribe this lecture audio accurately: subject '${recording.subject}', topic '${recording.title}'"),
                        Part(inlineData = InlineData(mimeType = mimeType, data = base64Audio))
                    )
                )
            ),
            generationConfig = GenerationConfig(
                responseMimeType = "text/plain",
                temperature = 0.1f
            )
        )

        onProgress("Transcribing spoken lecture audio to verbatim text...")
        val response = GeminiClient.apiService.generateContent(apiKey, request)
        val transcriptText = response.candidates?.firstOrNull()?.content?.parts?.firstOrNull()?.text?.trim()

        if (transcriptText.isNullOrBlank()) {
            throw IllegalStateException("Transcription failed: Gemini returned an empty transcript. Speech could not be detected.")
        }

        val transcriptId = "tr_" + UUID.randomUUID().toString().take(8)
        return@withContext Transcript(
            id = transcriptId,
            recordingId = recording.id,
            text = transcriptText,
            language = "en",
            durationSeconds = recording.durationSeconds,
            createdAt = System.currentTimeMillis(),
            status = TranscriptStatus.COMPLETED,
            isEdited = false,
            originalTextRef = null,
            isDemo = recording.isDemo
        )
    }
}
